ledger/sandbox/src/test/suite/scala/com/digitalasset/platform/resources/ResourceOwnerSpec.scala

```diff
@@ -0,0 +1,596 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.platform.resources
+
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.{Executors, RejectedExecutionException}
+import java.util.{Timer, TimerTask}
+
+import akka.actor.{Actor, ActorSystem, Props}
+import akka.stream.Materializer
+import akka.stream.scaladsl.{Keep, Sink, Source}
+import akka.{Done, NotUsed}
+import com.digitalasset.platform.resources.ResourceOwnerSpec._
+import org.scalatest.{AsyncWordSpec, Matchers}
+
+import scala.collection.mutable
+import scala.concurrent.duration.{DurationInt, FiniteDuration}
+import scala.concurrent.{ExecutionContext, Future, Promise}
+import scala.util.{Failure, Success}
+
+class ResourceOwnerSpec extends AsyncWordSpec with Matchers {
+  "a resource owner" should {
+    "acquire and release a resource" in {
+      val owner = TestResourceOwner(42)
+      owner.hasBeenAcquired should be(false)
+
+      val resource = for {
+        value <- owner.acquire()
+      } yield {
+        owner.hasBeenAcquired should be(true)
+        value should be(42)
+      }
+
+      for {
+        _ <- resource.asFuture
+        _ <- resource.release()
+      } yield {
+        owner.hasBeenAcquired should be(false)
+      }
+    }
+
+    "release all sub-resources when released" in {
+      val ownerA = TestResourceOwner(1)
+      val ownerB = TestResourceOwner("two")
+
+      val resource = for {
+        _ <- ownerA.acquire()
+        _ <- ownerB.acquire()
+      } yield {
+        ownerA.hasBeenAcquired should be(true)
+        ownerB.hasBeenAcquired should be(true)
+        ()
+      }
+
+      for {
+        _ <- resource.asFuture
+        _ <- resource.release()
+      } yield {
+        ownerA.hasBeenAcquired should be(false)
+        ownerB.hasBeenAcquired should be(false)
+      }
+    }
+
+    "only release once" in {
+      val ownerA = TestResourceOwner(7)
+      val ownerB = TestResourceOwner("eight")
+
+      val resource = for {
+        _ <- ownerA.acquire()
+        _ <- ownerB.acquire()
+      } yield ()
+
+      for {
+        _ <- resource.asFuture
+        _ <- resource.release()
+        _ <- resource.release() // will throw an exception if it actually releases twice
```

Comment:
So why does calling `release` twice not result in an exception? Probably better document the behavior and who has to do what with some comments. Adding them to the tests as well would be good, so that it's clear what is the runtime behavior we expect and the component involved in the test.

Reply:
It's because releases can and should be idempotent. I'll update the test name and comment to clarify.